File: packages/twilio_flutter_conversations/lib/src/client/connect_session_guard.dart
/// Guards stale [TwilioConversationsClient.connect] completions on the Dart side.
class ConnectSessionGuard {
  var _generation = 0;

  /// Starts a connect attempt; returns a token compared when the host call completes.
  int beginConnect() => _generation;

  /// Invalidates in-flight connect attempts (disconnect, dispose).
  void invalidate() {
    _generation += 1;
  }

  /// Returns true when [token] is still the active connect generation.
  bool isActive(int token) => token == _generation;
}

File: packages/twilio_flutter_conversations/android/src/main/kotlin/com/twilioflutter/conversations/bridge/ConnectSessionGuard.kt
package com.twilioflutter.conversations.bridge

/** Result of completing an asynchronous native [connect] callback. */
enum class ConnectCompletion {
  /** Connect result applied; client may be stored. */
  APPLIED,

  /** Connect was superseded; caller must shut down the SDK client and ignore storage. */
  STALE,
}

/**
 * Guards overlapping and stale Twilio client creation callbacks.
 *
 * Used by [ConversationsBridge] so disconnect/dispose during an in-flight connect
 * cannot leak a native client.
 */
class ConnectSessionGuard {
  private var generation: Int = 0
  private var isConnectInFlight: Boolean = false

  /** Returns the generation token when connect starts, or null if connect is already in flight. */
  fun beginConnect(): Int? {
    if (isConnectInFlight) {
      return null
    }
    isConnectInFlight = true
    return generation
  }

  /** Invalidates in-flight connect attempts (disconnect, dispose). */
  fun invalidatePendingConnect() {
    generation += 1
    isConnectInFlight = false
  }

  /** Completes a connect callback for [requestGeneration]. */
  fun completeConnect(requestGeneration: Int): ConnectCompletion {
    if (requestGeneration != generation) {
      return ConnectCompletion.STALE
    }
    isConnectInFlight = false
    return ConnectCompletion.APPLIED
  }

  /** Whether a connect call is currently in flight. */
  fun isConnectInFlight(): Boolean = isConnectInFlight
}
